{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.SeqUtils import GC\n",
    "from IPython.display import clear_output\n",
    "import os\n",
    "\n",
    "\n",
    "#Parameters to modify\n",
    "pmin = 20\n",
    "pmax = 25\n",
    "\n",
    "gc_min = 40\n",
    "gc_max = 60\n",
    "\n",
    "target_tm = 37\n",
    "\n",
    "min_amp_length = 150\n",
    "max_amp_length = 200\n",
    "\n",
    "dGpercentile = 75 #0-100\n",
    "\n",
    "primer_dimmer_threshold = 0.05 #0-1\n",
    "\n",
    "#Program variables\n",
    "v = []\n",
    "table = []\n",
    "buff = []\n",
    "final = []\n",
    "amplicons = []\n",
    "buff2 = []\n",
    "\n",
    "# OPENING CONSENSUS, REMOVING GAPS \n",
    "with open(\"Variation.txt\", 'r') as variation:\n",
    "    with open(\"consensusSeq.txt\", 'r') as infile:\n",
    "        s = SeqIO.read(infile, \"fasta\")\n",
    "        print(\"Sequence %s found.\" % (s.id))\n",
    "\n",
    "        for nc, var in zip(s.seq, variation.readlines()):\n",
    "            if nc != \"-\":\n",
    "                buff.append(nc)\n",
    "                v.append(float(var.split(\"\\t\")[1].split(\"\\n\")[0]))\n",
    "\n",
    "                \n",
    "s.seq = Seq(''.join(buff), generic_dna)\n",
    "buff = []\n",
    "\n",
    "# PRIMER ANALYSIS\n",
    "for primer_len in range(pmin, pmax):\n",
    "    #RULE 0: Length\n",
    "    for si in range(len(s) - primer_len + 1):\n",
    "\n",
    "        primer = s[si: si + primer_len] \n",
    "\n",
    "#           RULE 1: NO GAPS IN THE CONSENSOUS\n",
    "        if not \"-\" in primer:\n",
    "\n",
    "#               RULE 2: mean variation per nucleotide less than 1 & maximun variation less than 100\n",
    "            vm = sum(v[si: si + primer_len])/primer_len\n",
    "            if vm < 10 and max(v[si: si + primer_len]) < 100:\n",
    "\n",
    "#                   RULE 3: GC CONTENT\n",
    "                gc = GC(primer.seq)\n",
    "                if gc < gc_max and gc > gc_min:\n",
    "\n",
    "#                       RULE 4: Tm. ASSUMPTIONS 50 nM primer, 50 mM Na+, and pH 7.0 (http://www.lbgi.fr/MeltingTemperatureBioPhp.php)\n",
    "                    if primer_len > 13:\n",
    "                        tm = 64.9 + 41*(primer.seq.count(\"g\") + primer.seq.count(\"c\") - 16.4)/(primer.seq.count(\"g\") + primer.seq.count(\"c\") + primer.seq.count(\"a\") + primer.seq.count(\"t\"))\n",
    "                    else:\n",
    "                        tm = 2*(primer.seq.count(\"a\") + primer.seq.count(\"t\")) + 4*(primer.seq.count(\"g\") + primer.seq.count(\"c\"))\n",
    "                    if tm > target_tm:\n",
    "\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "#                           RULE 5: No homopolymers of length 4\n",
    "                        hp = False\n",
    "                        for i in range(primer_len - 3):   \n",
    "                            nc = primer.seq[i: i + 4]\n",
    "                            if (nc[0:2] == nc[2:4]):\n",
    "                                hp = True\n",
    "                            \n",
    "                        if not hp:\n",
    "                            table.append([si, primer.seq, gc, tm, vm])\n",
    "\n",
    "\n",
    "# PAIRING PRIMERS // TO-DO, PRIMERS WITH RELATIVELY THE SAME Tm\n",
    "for f in table:\n",
    "    for b in table:\n",
    "        d = b[0]-(f[0]+len(f[1]))\n",
    "        if d >  min_amp_length and d <  max_amp_length:\n",
    "            buff.append([f[1],b[1].reverse_complement(),f[0],b[0],s[f[0]:b[0]+len(b[1])]])\n",
    "        elif d > max_amp_length:\n",
    "            break\n",
    "amplicons = buff\n",
    "buff = []\n",
    "\n",
    "print(\"4 first rules and possible amplicons computed. Selected a total of %i amplicons.\\n\\nComputing dG pre-filtering (LinearFold) analysis...\" % (len(amplicons)))\n",
    "\n",
    "# RULE 5: LinearFold dG filtering\n",
    "for n,i in zip(amplicons, range(0,len(amplicons))):\n",
    "    out=os.popen('echo %s | /home/franquero/Documents/LinearFold/linearfold -V' % n[4].seq).read()\n",
    "    s = float(out.split(\" (\")[1].split(\")\")[0])\n",
    "    buff.append(n + [s])\n",
    "    clear_output(wait=True)\n",
    "    print(\"\\nProcessed %i amplicons.\\n\" % (i))\n",
    "amplicons = buff\n",
    "buff = []\n",
    "\n",
    "#Selectiong the ones inside of the chosen dG percentile\n",
    "for n in amplicons:\n",
    "    buff.append(n[5])\n",
    "    \n",
    "Glimit = np.percentile(buff,dGpercentile) \n",
    "buff = []\n",
    "print(\"Percentil %i of dG in the primer sample is %i.\\n\" % (dGpercentile, Glimit))\n",
    "\n",
    "for n in amplicons:\n",
    "    if n[5] > Glimit:\n",
    "        buff.append(n)\n",
    "        \n",
    "amplicons = buff\n",
    "buff = []\n",
    "\n",
    "print(\"dG pre-filtering computed. Selected a total of %i amplicons.\\n\\nComputing primer dimmer analysis...\" % (len(amplicons)))\n",
    "\n",
    "# RULE 6: NUPACK // PRIMER DIMER & REGION NOT STRUCTURED\n",
    "        \n",
    "\n",
    "for amp in amplicons:\n",
    "    with open (\"temp.in\", \"w\") as infile:\n",
    "        infile.write(\"2\" + '\\n' + ('\\n'.join([str(amp[0]) ,str(amp[1])])) + '\\n' + \"1 2\" + '\\n')\n",
    "        infile.close()\n",
    "\n",
    "    cmd_input = str.encode(\"temp\")\n",
    "    p = sub.Popen([\"pairs\", \"-multi\",\"-material\", \"dna\", \"temp\"], stdin=sub.PIPE, stdout=sub.PIPE, stderr=sub.STDOUT)\n",
    "    stdout = p.communicate()\n",
    "\n",
    "    with open(\"temp.ppairs\", \"r\") as temp:   \n",
    "        outputlines = temp.readlines()\n",
    "        temp.close()\n",
    "        os.remove(\"temp.ppairs\")\n",
    "        os.remove(\"temp.epairs\")\n",
    "        os.remove(\"temp.in\")\n",
    "\n",
    "    for line in outputlines:\n",
    "        if len(line.split(\"\\t\")) > 1 and line[0] != '\\n' in line:\n",
    "            if len(buff) == 0:\n",
    "                buff = np.array([[float(i) for i in line.split(\"\\t\")[0:3]]])\n",
    "            else:\n",
    "                buff = np.append(buff, [[float(i) for i in line.split(\"\\t\")[0:3]]],0)\n",
    "\n",
    "    score = 0\n",
    "    for n in buff:\n",
    "        nc = n[0] - len(amp[0]) \n",
    "        if -10 < nc <= 0:\n",
    "            nc = n[1] - len(amp[0]) \n",
    "            if 0 < nc <= 10:\n",
    "                k = 1 + ((5 - nc) * 0.2)\n",
    "                if score == 0:\n",
    "                    score = (score + (k * n[2]))\n",
    "                else:\n",
    "                    score = (score + (k * n[2]))/2\n",
    "                #print(str(k) + \" \" + str(n[2]) + \" \" + str(score))\n",
    "\n",
    "        nc = n[0] - len(amp[0])        \n",
    "        if 0 < nc <= 10:\n",
    "            nc = n[1] - len(amp[0]) \n",
    "            if -10 < nc <= 0:\n",
    "                k = 1 + ((5 + nc) * 0.2)\n",
    "                if score == 0:\n",
    "                    score = (score + (k * n[2]))\n",
    "                else:\n",
    "                    score = (score + (k * n[2]))/2\n",
    "                #print(str(k) + \" \" + str(n[2]) + \" \" + str(score))\n",
    "\n",
    "    if  score < primer_dimmer_threshold:\n",
    "            amp.append(score)\n",
    "            buff2.append(amp)\n",
    "        \n",
    "    buff = []\n",
    "    \n",
    "\n",
    "    \n",
    "amplicons = buff2\n",
    "buff2 = []\n",
    "\n",
    "\n",
    "\n",
    "print(\"Primer dimmer analysis computed. Selected a total of %i amplicons.\\n\\nComputing amplicon structure analysis...\" % (len(amplicons)))\n",
    "\n",
    "#NuPack amplicon analysis\n",
    "\n",
    "#FINAL LIST OF PRIMERS FOR BLAST\n",
    "for n in table:\n",
    "    include = False\n",
    "    for amp in amplicons:\n",
    "        if n[1] == amp[0] or n[1].reverse_complement() == amp[1]:\n",
    "            include = True\n",
    "    if include:\n",
    "        buff.append(n)\n",
    "        \n",
    "table = buff\n",
    "buff = []\n",
    "\n",
    "print(\"%i total primers for %i amplicons.\\n\\nComputing BLASTn analysis.\" % (len(table),len(amplicons)))\n",
    "\n",
    "#RULE 7: BLAST\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
