{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "consensus = []\n",
    "nc_range = {\n",
    "    \"F3\":0,\n",
    "    \"B3\":0,\n",
    "    \"FIP\":0,\n",
    "    \"rFIP\":0,\n",
    "    \"BIP\":0,\n",
    "    \"rBIP\":0\n",
    "}\n",
    "\n",
    "range_variation = []\n",
    "plocation = []\n",
    "primers={\n",
    "    \"F3\":\"\",\n",
    "    \"B3\":\"\",\n",
    "    \"FIP\":\"\",\n",
    "    \"rFIP\":\"\",\n",
    "    \"BIP\":\"\",\n",
    "    \"rBIP\":\"\"\n",
    "}\n",
    "\n",
    "\n",
    "print(\"________________ALIGNING PRIMERS_________________\")\n",
    "with open(\"consensusSeq.txt\", 'r') as infile:\n",
    "    for record in SeqIO.parse(infile, \"fasta\"):\n",
    "        consensus = str(record.seq).upper()\n",
    "    with open(\"primers.txt\", 'r') as primersIO:\n",
    "            for primersIO in SeqIO.parse(primersIO, \"fasta\"):\n",
    "                if \"F3\" in primersIO.id:\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-1.5, -10,-2)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"F3\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"F3\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"B3\" in primersIO.id:\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-1.5, -10,-2)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"B3\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"B3\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"FIP\" in primersIO.id:\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-1.5, -10,-2)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"FIP\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"FIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-1.5, -10,-2)\n",
    "                    print(\"----------REVERSE COMPLEMENT OF \" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"rFIP\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"rFIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"BIP\" in primersIO.id:\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-1.5, -10,-2)\n",
    "                    print(\"----------REVERSE COMPLEMENT OF \" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"rBIP\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"rBIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-1.5, -10,-2)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    nc_range[\"BIP\"] = format_alignment(*alignments[0]).split(\" \")[0]\n",
    "                    primers[\"BIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "            \n",
    "print(\"PRIMERS RANGE:\" + str(nc_range[\"F3\"]) + \" - \" + str(nc_range[\"B3\"]))\n",
    "\n",
    "print(\"________________COMPUTING VARIATION MATRIX_________________\")\n",
    "\n",
    "with open(\"variation.txt\", 'r') as infile:\n",
    "    \n",
    "    nc_variation = infile.readline()\n",
    "    while not nc_range[\"F3\"] in nc_variation:\n",
    "        nc_variation = infile.readline()\n",
    "        \n",
    "    range_variation.append(float(nc_variation.split(\"\\t\")[1].split(\"\\n\")[0]))\n",
    "    \n",
    "    while not nc_range[\"B3\"] in nc_variation:\n",
    "        nc_variation = infile.readline()\n",
    "        range_variation.append(float(nc_variation.split(\"\\t\")[1].split(\"\\n\")[0]))\n",
    "    \n",
    "    buff = []\n",
    "    i = 0\n",
    "\n",
    "    for n in range_variation:\n",
    "            #calculate variation\n",
    "            if n < 10:\n",
    "                buff.append(0)\n",
    "            elif n < 80:\n",
    "                buff.append(1)\n",
    "            elif n < 800:\n",
    "                buff.append(2)\n",
    "            else:\n",
    "                buff.append(3)\n",
    "            \n",
    "            #Do this region have primers?\n",
    "            if i < len(str(primers[\"F3\"])): #F3\n",
    "                plocation.append(1)\n",
    "            elif i > (len(np.array(range_variation)) - len(str(primers[\"B3\"]))):\n",
    "                plocation.append(9)\n",
    "            elif i > (int(nc_range[\"FIP\"])-int(nc_range[\"F3\"])) and i < ((int(nc_range[\"FIP\"])-int(nc_range[\"F3\"])) + len(str(primers[\"FIP\"]))): \n",
    "                plocation.append(3)\n",
    "            elif i > (int(nc_range[\"rBIP\"])-int(nc_range[\"F3\"])) and i < ((int(nc_range[\"rBIP\"])-int(nc_range[\"F3\"])) + len(str(primers[\"rBIP\"]))): \n",
    "                plocation.append(6) \n",
    "            elif i < (int(nc_range[\"BIP\"])-int(nc_range[\"F3\"])) and i > ((int(nc_range[\"BIP\"])-int(nc_range[\"F3\"])) - len(str(primers[\"BIP\"]))): \n",
    "                plocation.append(6)\n",
    "            elif i < (int(nc_range[\"rFIP\"])-int(nc_range[\"F3\"])) and i > ((int(nc_range[\"rFIP\"])-int(nc_range[\"F3\"])) - len(str(primers[\"rFIP\"]))): \n",
    "                plocation.append(3)\n",
    "            else:\n",
    "                plocation.append(0)\n",
    "                \n",
    "            i = i + 1\n",
    "                \n",
    "    range_variation = buff\n",
    "    \n",
    "    rv = np.array([range_variation])\n",
    "    rv = np.vstack((rv, np.array([plocation])))\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(20, 0.8))\n",
    "    im = ax.imshow(rv,aspect='auto',extent=[int(nc_range[\"F3\"]),int(nc_range[\"B3\"]),0,1])\n",
    "    \n",
    "    \n",
    "    plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
